docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,130 @@
+	// group specified by the groupUID
+	ExtractMembers(groupUID string) (members []*ldap.Entry, err error)
+	// ExtractAnnotations returns the mapping of annotation keys to values that the OpenShift group
+	// representing the LDAP group specified by the groupUID should have
+	ExtractAnnotations(groupUID string) (annotations map[string]string, err error)
+
+	LDAPGroupNameMapper
+}
+
+// LDAPGroupNameMapper maps a groupUID representing an LDAP group to the OpenShift Group name for
+// the resource
+type LDAPGroupNameMapper interface {
+	GroupNameFor(groupUID string) (groupName string, err error)
+}
+```
+
+## Determining OpenShift `Identities` for LDAP Members
+This will be done in an identical way to the mapping for [LDAP password processing](https://github.com/openshift/origin/blob/master/pkg/auth/authenticator/password/ldappassword/ldap.go#L227-243) once the requisite data is collected.
+
+```go
+// LDAPIdentityFactory creates Identites for LDAP user entries.
+type LDAPUserIdentityFactory interface {
+	IdentityFor(user *ldap.Entry) (identity userapi.Identity, err error)
+}
+```
+
+## Determining OpenShift `Users` for OpenShift `Identities`
+A new `UserIdentityMapper` will need to be created: the `DeterministicUserIdentityMapper`, which will assume that the unique identifier retrieved from the LDAP response and integrated into the OpenShift `Identity` will be deterministicically mapped to an OpenShift `User`. This is required as otherwise there is no deterministic mapping from the `Identity` created and any OpenShift `User`.
+
+```go
+// DeterministicUserIdentityMapper is a UserIdentityMapper that forces a deterministic mapping from
+// an Identity to a User.
+type DeterministicUserIdentityMapper struct {
+  // AllowIdentityCollisions determines if this UserIdentityMapper is allowed to make a mapping
+  // between an Identity provided by LDAP and a User that already is mapped to another Identity.
+  AllowIdentityCollisions bool
+}
+
+func (m *DeterministicUserIdentityMapper) UserFor(identity authapi.UserIdentityInfo) (user user.Info,
+	err error) {
+}
+```
+
+## Creating OpenShift `Groups`
+1. Find the OpenShift `Group` or start one if it does not exist
+* Populate `Annotations` and `Labels`
+* Overwrite `Users` list
+* Replace current OpenShift `Group` or create it
+
+## Exposed Options for `oadm sync-groups`
+The sync command will be exposed as `oadm sync-groups [<openshift-group-name>...] --config=<location>` and invoked like:
+* `oadm sync-groups --all-openshift --config=/etc/openshift/ldap-sync.yaml`
+* `oadm sync-groups <names> --config=/etc/openshift/ldap-sync.yaml`
+* `oadm sync-groups --all-ldap --config=/etc/openshift/ldap-sync.yaml`
+* `oadm sync-groups --whitelist-ldap --config=/etc/openshift/ldap-sync.yaml`
+* `oadm sync-groups --prune --config=/etc/openshift/ldap-sync.yaml`
+
+`--config=/path/to/file` determines where the config `yaml` is located.
+
+`--all-openshift` would specify that all OpenShift `Groups` need to be synced
```

Comment:
All OpenShift `Groups`. This would allow for existing records to be matched with / pick up LDAP sync. If that's not a valid use-case, and we want to fail when an LDAP group matches a OpenShift `Group` that doesn't currently have LDAP sync annotations, then this will change or be removed.